File: src/main/java/dev/adamko/gildedrose/java/item/ItemTemplateProvider.java
package dev.adamko.gildedrose.java.item;

import dev.adamko.gildedrose.item.ItemTemplateProvider.Companion;
import kotlin.text.StringsKt;


/**
 * A Java implementation of {@link dev.adamko.gildedrose.item.ItemTemplateProvider}
 *
 * @see dev.adamko.gildedrose.item.ItemTemplateProvider
 */
@FunctionalInterface
public interface ItemTemplateProvider {

  ItemTemplate getTemplate(ItemState item);

  /**
   * A Java implementation of {@link Companion#getDefaultTemplateProvider()}
   *
   * @see Companion#getDefaultTemplateProvider()
   */
  ItemTemplateProvider defaultTemplateProvider = item -> {

    // conjured
    if (StringsKt.startsWith(item.name(), "Conjured", false)) {
      return ItemTemplates.conjuredItem;
    }
    // tickets
    else if ("Backstage passes to a TAFKAL80ETC concert".equals(item.name())) {
      return ItemTemplates.ticketItem;
    }
    // legendary
    else if ("Sulfuras, Hand of Ragnaros".equals(item.name())) {
      return ItemTemplates.legendaryItem;
    }
    // aged
    else if ("Aged Brie".equals(item.name())) {
      return ItemTemplates.agedItem;
    }
    // regular
    else {
      return ItemTemplates.regularItem;
    }

  };

}

File: src/main/kotlin/dev/adamko/gildedrose/item/ItemTemplateProvider.kt
package dev.adamko.gildedrose.item


/**
 * Provides an [ItemTemplate] for a given [ItemState].
 */
fun interface ItemTemplateProvider {

  fun getTemplate(item: ItemState): ItemTemplate

  companion object {

    val defaultTemplateProvider = ItemTemplateProvider {
      it.name.run {
        when {
          // conjured
          startsWith("Conjured")                              -> ItemTemplates.conjuredItem
          // tickets
          equals("Backstage passes to a TAFKAL80ETC concert") -> ItemTemplates.ticketItem
          // legendary
          equals("Sulfuras, Hand of Ragnaros")                -> ItemTemplates.legendaryItem
          // aged
          equals("Aged Brie")                                 -> ItemTemplates.agedItem
          // regular
          else                                                -> ItemTemplates.regularItem
        }
      }
    }
  }
}
